### PR TITLE
Add Textbox Character Set Validation

### LINKED ISSUE #22: Textbox Character Set Validation
Geometry Dash textboxes feature a limited character space. This is usually only includes alphanumeric characters, but may occasionally allow extra characters (such as email allowing `@` and passwords including more characters).

Enforcing this would be beneficial as the Geometry Dash response format does **not** support escaping characters. This means that currently, a level with a `|` in its name can genuinely completely prevent the game from reading the response.

Additionally, would help against botting.

An ideal solution would leverage pydantic's parsing, perhaps an implementation similar to the `Base64String` validator. Should be efficient as it will likely be used on many endpoints. https://github.com/RealistikDash/RealistikGDPS/blob/d767f4083241f9fbd40771bba4fbfa6c4a2cce62/rgdps/common/validators.py#L9-L30

### PR DESCRIPTION
**Pull request by lenforiee**
As mentioned in #22 illegal characters can cause huge trouble eg. client not reading the response, there is client-side validation but this can still be exploited by sending a faked request with curl or Postman. The goal of this PR is to validate the data sent to the server (in vulnerable places)

### What's in this PR:
- [x] Custom text box string validation for username on login/register request and level name on submit level request
- [x] Custom social media string validation when adding your youtube, twitter or twitch to your account

### CODE REVIEW

File: rgdps/common/validators.py [hunk 1]
```diff
@@ -28,3 +34,31 @@ def _validate(cls, value: Any, _: core_schema.ValidationInfo) -> Base64String:
...
+        if not isinstance(value, (str, bytes)):
+            raise TypeError("Value must be str or bytes")
+
+        if isinstance(value, bytes):
+            value = value.decode()
+
+        # Value needs to be: stripped and alphanumeric.
+        value = value.strip()
+
+        if not value.isalnum():
```
Comment (RealistikDash) [inline rgdps/common/validators.py]: image
The function `str.isalnum()` does not allow spaces, which some Geometry Dash textboxes allow (such as level name).

**Review by RealistikDash — CHANGES_REQUESTED** [review]

File: rgdps/common/validators.py (hunk 1 above)
Comment (lenforiee) [reply rgdps/common/validators.py]: should be fixed in 08c940a

File: rgdps/common/validators.py [hunk 2]
```diff
@@ -28,3 +35,59 @@ def _validate(cls, value: Any, _: core_schema.ValidationInfo) -> Base64String:
...
+        if not isinstance(value, (str, bytes)):
+            raise TypeError("Value must be str or bytes")
+
+        if isinstance(value, bytes):
+            value = value.decode()
+
+        # Value needs to be: stripped and alphanumeric.
+        value = value.strip()
+
+        if not re.match(r"^[a-zA-Z0-9 ]+$", value):
```
Comment (RealistikDash) [inline rgdps/common/validators.py]: Would pre-compining the regex be beneficial?

File: rgdps/common/validators.py [hunk 3]
```diff
@@ -28,3 +35,59 @@ def _validate(cls, value: Any, _: core_schema.ValidationInfo) -> Base64String:
...
+        if not isinstance(value, (str, bytes)):
+            raise TypeError("Value must be str or bytes")
+
+        if isinstance(value, bytes):
+            value = value.decode()
+
+        # Value needs to be: stripped and alphanumeric + it can contain `_`, `-`, `.`, and `'`.
+        value = value.strip()
+
+        if not re.match(r"^[\w\-.' ]+$", value):
```
Comment (RealistikDash) [inline rgdps/common/validators.py]: Similar to above comment.

**Review by RealistikDash — COMMENTED** [review]
Regex compilation

File: rgdps/common/validators.py (hunk 2 above)
Comment (cmyui) [reply rgdps/common/validators.py]: i think you can just use https://docs.python.org/3/library/stdtypes.html#str.isalnum

File: rgdps/common/validators.py (hunk 3 above)
Comment (cmyui) [inline rgdps/common/validators.py:90]:
```suggestion
        if not value.isalnum():
```

File: rgdps/common/validators.py [hunk 4]
```diff
@@ -1,16 +1,23 @@
 from __future__ import annotations
 
+import re
```
Comment (cmyui) [inline rgdps/common/validators.py:3]:
```suggestion
```

File: rgdps/common/validators.py (hunk 2 above)
Comment (cmyui) [reply rgdps/common/validators.py]: oo actually no because of the spaces

Comment (RealistikDash) [reply rgdps/common/validators.py]: Was `str.isalnum` prior, addressed in previous review.

**Review by RealistikDash — APPROVED** [review]
Looks good!

File: rgdps/common/validators.py [hunk 5]
```diff
@@ -87,7 +90,7 @@ def _validate(cls, value: Any, _: core_schema.ValidationInfo) -> TextBoxString:
         # Value needs to be: stripped and alphanumeric + it can contain `_`, `-`, `.`, and `'`.
         value = value.strip()
 
-        if not re.match(r"^[\w\-.' ]+$", value):
+        if not SOCIAL_MEDIA_REGEX.match(value):
             raise ValueError(f"Input contains illegal characters")
```
Comment (cmyui) [inline rgdps/common/validators.py:94]: fstrings?

Comment (RealistikDash) [reply rgdps/common/validators.py:94]: Good catch!

**Pull request merged**